numba/tests/test_runtests.py

```diff
@@ -119,34 +125,66 @@ def get_count(arg_list):
             excluded = get_count(['--exclude-tags=%s' % tag, 'numba.tests'])
             self.assertEqual(total, included + excluded)
 
-    def test_check_slice(self):
+    def test_check_shard(self):
         tmpAll = self.get_testsuite_listing([])
         tmp1 = self.get_testsuite_listing(['-j', '0:2'])
         tmp2 = self.get_testsuite_listing(['-j', '1:2'])
-        lAll = {x for x in tmpAll if x.startswith('numba.')}
-        l1 = {x for x in tmp1 if x.startswith('numba.')}
-        l2 = {x for x in tmp2 if x.startswith('numba.')}
+
+        lAll = set(self._get_numba_tests_from_listing(tmpAll))
+        l1 = set(self._get_numba_tests_from_listing(tmp1))
+        l2 = set(self._get_numba_tests_from_listing(tmp2))
+
         # The difference between two adjacent shards should be less than 5% of
         # the total
         self.assertLess(abs(len(l2) - len(l1)), len(lAll) / 20)
         self.assertLess(len(l1), len(lAll))
         self.assertLess(len(l2), len(lAll))
-        # Test should always run
-        self.assertIn("numba.tests.test_api.TestNumbaModule.test_numba_module",
-                      l1)
-        self.assertIn("numba.tests.test_api.TestNumbaModule.test_numba_module",
-                      l2)
-
-    def test_check_slicing_equivalent(self):
-        def filter_test(xs):
-            return {x for x in xs if x.startswith('numba.')}
-        full = filter_test(self.get_testsuite_listing([]))
-        sliced = set()
+
+    def test_check_sharding_equivalent(self):
+        # get some shards
+        sharded = list()
         for i in range(3):
             subset = self.get_testsuite_listing(['-j', '{}:3'.format(i)])
-            sliced.update(filter_test(subset))
-        # The tests must be equivalent
-        self.assertEqual(full, sliced)
+            slist = [*self._get_numba_tests_from_listing(subset)]
+            # there's at least 100 tests in a shard of roughly 1/3!
+            self.assertGreater(len(slist), 100)
```

Comment:
This seems fragile to put in this value.

Reply:
The intent here is to make sure that there's not an "accident" whereby the sharding logic has produced an unreasonable quantity of tests in a shard given the size of the test suite. Having seen a few mistakes made through the history of the test listing logic, they tend to manifest as there being be zero or just a few tests listed. As a result I chose 100 as a minimum based on this and as there are 10k+ tests in total it's unlikely a ~1/3 split would ever be less than 100. Also note there are other baked in constants (see `check_all` above which makes sure there are more than 5k tests listed).

Do you think there might be something less "fragile" to check or some other way to assert similar? I suppose querying the size of the test suite and taking a percentage of 1/3 of that might be another way?

Reply:
Take the total and divide by 3. Then it doesn't matter if the test count changes. Or check that the slices are approximately the same size. If they sum to the total and are about equal, then the actual number doesn't matter.

Reply:
Maybe it's best to just remove this assertion. In this module:
* `test_check_shard` asserts that shards are "approximately the same size".
* `test_default`/`test_all` both check that the test suite is being listed as a sensible size (>5000 tests).
* this test checks that the shard sizes add up to the total which is the same as obtained in `test_default`.

I think this renders the assertion under discussion well intentioned but superfluous.

Reply:
The above is done in b4081c1.